futures/examples/async_context.rs

```diff
@@ -0,0 +1,62 @@
+#![feature(async_await, futures_api, await_macro, pin, arbitrary_self_types)]
+extern crate futures;
+extern crate rand;
+
+use futures::{
+    executor //standard executors to provide a context for futures and streams
+};
+
+fn main() {
+    //
+    // async implements future, so will be passed a std::task::Context when
+    // future::poll is invoked. Once the context is passed, it will then be
+    // available within the async block for additional async/await operations.
+    //
+    // This is approximately equal to having a thread local context that is
+    // available within the future, and given to additional async/await blocks.
+    //
+    // ```text
+    //   static mut TASK_CONTEXT: Option<NonNull<std::task::Context>> = None;
+    //   let polled = unsafe {
+    //     let mut cx = TASK_CONTEXT.take().unwrap();
+    // ```
+    //
+    let fut_result = async {
+        //
+        // Once inside the async block and its std::task::Context, we can use
+        // additional async or await blocks, utilizing the std::task::Context
+        // that this async block has maintained.
+        //
+        // For async blocks, the std::task::Context of the parent block will be
+        // passed via future::poll to the child async blocks.
+        //
+        let fut_value = async {
+            rand::random::<i32>()
+        };
+
+        //
+        // For await! macros, the std::task::Context of the parent async block
```

Comment:
`await!` is currently a macro. It will be a compiler-builtin. But that's technical. Probably best to call it an `await!` expression.